pkg/providers/snow/snow.go

```diff
@@ -176,19 +183,47 @@ func (p *snowProvider) GetInfrastructureBundle(clusterSpec *cluster.Spec) *types
 }
 
 func (p *snowProvider) DatacenterConfig() providers.DatacenterConfig {
-	return nil
+	return p.datacenterConfig
 }
 
 func (p *snowProvider) DatacenterResourceType() string {
-	return ""
+	return snowDatacenterResourceType
 }
 
 func (p *snowProvider) MachineResourceType() string {
-	return ""
+	return snowMachineResourceType
 }
 
 func (p *snowProvider) MachineConfigs() []providers.MachineConfig {
-	return nil
+	configs := make(map[string]providers.MachineConfig, len(p.machineConfigs))
+	controlPlaneMachineName := p.clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name
+	p.machineConfigs[controlPlaneMachineName].Annotations = map[string]string{p.clusterConfig.ControlPlaneAnnotation(): "true"}
+	if p.clusterConfig.IsManaged() {
+		p.machineConfigs[controlPlaneMachineName].SetManagedBy(p.clusterConfig.ManagedBy())
+	}
+	configs[controlPlaneMachineName] = p.machineConfigs[controlPlaneMachineName]
+
+	if p.clusterConfig.Spec.ExternalEtcdConfiguration != nil {
+		etcdMachineName := p.clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name
+		p.machineConfigs[etcdMachineName].Annotations = map[string]string{p.clusterConfig.EtcdAnnotation(): "true"}
+		if etcdMachineName != controlPlaneMachineName {
+			configs[etcdMachineName] = p.machineConfigs[etcdMachineName]
+			if p.clusterConfig.IsManaged() {
+				p.machineConfigs[etcdMachineName].SetManagedBy(p.clusterConfig.ManagedBy())
+			}
+		}
+	}
+
+	for _, workerNodeGroupConfiguration := range p.clusterConfig.Spec.WorkerNodeGroupConfigurations {
+		workerMachineName := workerNodeGroupConfiguration.MachineGroupRef.Name
+		if _, ok := configs[workerMachineName]; !ok {
+			configs[workerMachineName] = p.machineConfigs[workerMachineName]
+			if p.clusterConfig.IsManaged() {
+				p.machineConfigs[workerMachineName].SetManagedBy(p.clusterConfig.ManagedBy())
+			}
+		}
+	}
+	return providers.MachineConfigsMapToSlice(configs)
```

Comment:
can't we just use `p.machineConfigs` here instead of building `configs`?
It would avoid having this 3 sections of code that basically do the same thing